controlplane/kubeadm/internal/kubeadm_config_map.go

```diff
@@ -196,6 +202,83 @@ func (k *kubeadmConfig) UpdateCoreDNSImageInfo(repository, tag string) error {
 	return nil
 }
 
+// UpdateAPIServer sets the api server configuration to values set in `apiServer` in kubeadm config map.
+func (k *kubeadmConfig) UpdateAPIServer(apiServer kubeadmv1.APIServer) (bool, error) {
+	changed, err := k.updateClusterConfiguration(apiServer, apiServerKey)
+	if err != nil {
+		return false, errors.Wrap(err, "unable to update api server configuration in kubeadm config map")
+	}
+	return changed, nil
+}
+
+// UpdateControllerManager sets the controller manager configuration to values set in `controllerManager` in kubeadm config map.
+func (k *kubeadmConfig) UpdateControllerManager(controllerManager kubeadmv1.ControlPlaneComponent) (bool, error) {
+	changed, err := k.updateClusterConfiguration(controllerManager, controllerManagerKey)
+	if err != nil {
+		return false, errors.Wrap(err, "unable to update controller manager configuration in kubeadm config map")
+	}
+	return changed, nil
+}
+
+// UpdateScheduler sets the scheduler configuration to values set in `scheduler` in kubeadm config map.
+func (k *kubeadmConfig) UpdateScheduler(scheduler kubeadmv1.ControlPlaneComponent) (bool, error) {
+	changed, err := k.updateClusterConfiguration(scheduler, schedulerKey)
+	if err != nil {
+		return false, errors.Wrap(err, "unable to update scheduler configuration in kubeadm config map")
+	}
+	return changed, nil
+}
+
+// UpdateFeatureGates sets the feature gates to values set in `featureGates` in kubeadm config map.
+func (k *kubeadmConfig) UpdateFeatureGates(featureGates map[string]bool) (bool, error) {
+	changed, err := k.updateClusterConfiguration(featureGates, featureGatesKey)
+	if err != nil {
+		return false, errors.Wrap(err, "unable to update feature gates in kubeadm config map")
+	}
+	return changed, nil
+}
+
+// updateClusterConfiguration is a generic method to update any kubeadm ClusterConfiguration spec with custom types in the specified path.
+func (k *kubeadmConfig) updateClusterConfiguration(config interface{}, path ...string) (bool, error) {
+	data, ok := k.ConfigMap.Data[clusterConfigurationKey]
```

Comment:
should we test for k and ConfigMap not nil (might be overprotective...)

Reply:
`configMap` cannot be nil because otherwise it would fail [here](https://github.com/kubernetes-sigs/cluster-api/blob/0b16297abf41863dbae7ffda0f5fb06bbdd9015e/controlplane/kubeadm/internal/workload_cluster.go#L115) . Since we set `k` only if err == nil [here](https://github.com/kubernetes-sigs/cluster-api/blob/0b16297abf41863dbae7ffda0f5fb06bbdd9015e/controlplane/kubeadm/internal/workload_cluster.go#L142), it cannot be nil either. I think it might be redundant to do another nil check here.

Reply:
thanks for explanation